### PR TITLE
[Algorithm] Update DQN example

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

Reproduce known DQN benchmarks.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: examples/dqn/dqn_atari.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+#     env.append_transform(GrayScale())
+#     env.append_transform(Resize(84, 84))
+#     env.append_transform(CatFrames(N=4, dim=-3))
+#     env.append_transform(RewardSum())
+#     env.append_transform(StepCounter(max_steps=4500))
+#     if not is_test:
+#         env.append_transform(RewardClipping(-1, 1))
+#     env.append_transform(DoubleToFloat())
+#     # env.append_transform(VecNorm(in_keys=["pixels"]))
+#     return env
```
Comment (vmoens) [inline examples/dqn/dqn_atari.py:93]: What makes it crash? Anything we can fix?

File: examples/dqn/dqn_atari.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    proof_environment = make_env(env_name, frame_skip, device="cpu")
+    qvalue_module = make_dqn_modules_pixels(proof_environment)
+    del proof_environment
+    return qvalue_module
+
+
+@hydra.main(config_path=".", config_name="config_atari", version_base="1.1")
+def main(cfg: "DictConfig"):  # noqa: F821
+
+    device = "cpu" if not torch.cuda.is_available() else "cuda"
```
Comment (vmoens) [inline examples/dqn/dqn_atari.py:178]:
```suggestion
    device = "cpu" if not torch.cuda.device_count() else "cuda"
```

File: examples/dqn/dqn_atari.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+                    )
+                    reward = td_test["next", "episode_reward"][td_test["next", "done"]]
+                    test_rewards = np.append(test_rewards, reward.cpu().numpy())
+                    del td_test
+                logger.log_scalar("eval/reward", test_rewards.mean(), collected_frames)
+                model.train()
+
+        # update weights of the inference policy
+        collector.update_policy_weights_()
+        sampling_start = time.time()
```
Comment (vmoens) [inline examples/dqn/dqn_atari.py:329]: To keep the training loop short, we can put the testing in a separate function, as well as the calls to the logger

File: examples/dqn/dqn_cartpole.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
+"""
```
Comment (vmoens) [inline examples/dqn/dqn_cartpole.py:1]: missing headers

File: examples/dqn/dqn_cartpole.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    proof_environment = make_env(env_name, device="cpu")
+    qvalue_module = make_dqn_modules(proof_environment)
+    del proof_environment
+    return qvalue_module
+
+
+@hydra.main(config_path=".", config_name="config_cartpole", version_base="1.1")
+def main(cfg: "DictConfig"):  # noqa: F821
+
+    device = "cpu" if not torch.cuda.is_available() else "cuda"
```
Comment (vmoens) [inline examples/dqn/dqn_cartpole.py:82]:
```suggestion
    device = "cpu" if not torch.cuda.device_count() else "cuda"
```

File: examples/dqn/dqn_cartpole.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                    )
+                    reward = td_test["next", "episode_reward"][td_test["next", "done"]]
+                    test_rewards = np.append(test_rewards, reward.cpu().numpy())
+                    del td_test
+                logger.log_scalar("eval/reward", test_rewards.mean(), collected_frames)
+                model.train()
+
+        # update weights of the inference policy
+        collector.update_policy_weights_()
+        sampling_start = time.time()
```
Comment (vmoens) [inline examples/dqn/dqn_cartpole.py:219]: ditto, let's consider wiriting test and logging within dedicated funcs

**Review by vmoens — APPROVED** [review]
LGTM, some minor comments

File: examples/dqn/dqn_atari.py (hunk 1 above)
Comment (albertbou92) [reply examples/dqn/dqn_atari.py]: This it what makes it crash: https://github.com/pytorch/rl/issues/1518

File: setup.py [hunk 7]
```diff
@@ -186,7 +186,7 @@ def _main(argv):
     # tag = _run_cmd(["git", "describe", "--tags", "--exact-match", "@"])
 
     this_directory = Path(__file__).parent
-    long_description = (this_directory / "README.md").read_text()
```
Comment (vmoens) [inline setup.py:189]: interesting :)

File: README.md [hunk 8]
```diff
@@ -490,7 +490,7 @@ If you would like to contribute to new features, check our [call for contributio
 ## Examples, tutorials and demos
 
 A series of [examples](examples/) are provided with an illustrative purpose:
-- [DQN and Rainbow](examples/dqn/dqn.py)
```
Comment (vmoens) [inline README.md:493]: Perhaps we want to link to the parent directory?
We mention rainbow there: is there a plan to incorporate it too? Maybe as a separate script?

File: examples/dqn/dqn_cartpole.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    sampling_start = time.time()
+
+    for data in collector:
+
+        log_info = {}
+        sampling_time = time.time() - sampling_start
+        pbar.update(data.numel())
+        data = data.reshape(-1)
+        current_frames = data.numel()
+        replay_buffer.extend(data.to(device))
```
Comment (vmoens) [inline examples/dqn/dqn_cartpole.py:102]: seems weird that we are sending the data to device before putting it in the replay buffer.
Usually the buffer is on disk/ram and the data on cuda, not the other way around

File: examples/dqn/dqn_cartpole.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        training_start = time.time()
+        for j in range(num_updates):
+            sampled_tensordict = replay_buffer.sample(batch_size)
+            loss_td = loss_module(sampled_tensordict)
+            q_loss = loss_td["loss"]
+            optimizer.zero_grad()
+            q_loss.backward()
+            optimizer.step()
+            target_net_updater.step()
+            q_losses[j] = loss_td.select("loss").detach()
```
Comment (vmoens) [inline examples/dqn/dqn_cartpole.py:135]: Could/should we log more things?

File: examples/dqn/dqn_atari.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+        init_random_frames=init_random_frames,
+    )
+
+    # Create the replay buffer
+    replay_buffer = TensorDictReplayBuffer(
+        pin_memory=False,
+        prefetch=3,
+        storage=LazyMemmapStorage(
+            max_size=cfg.buffer.buffer_size,
+            scratch_dir="/tmp/",
```
Comment (vmoens) [inline examples/dqn/dqn_atari.py:65]: the scratch dir should be unique, see how we did it in td3

**Review by vmoens — COMMENTED** [review]
Thanks for this! Great work, very clean!

I mainly commented on cartpole but most comments also apply to atari.
Let me know if I can help with anything!

File: examples/dqn/dqn_cartpole.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        episode_rewards = data["next", "episode_reward"][data["next", "done"]]
+        if len(episode_rewards) > 0:
+            episode_length = data["next", "step_count"][data["next", "done"]]
+            log_info.update(
+                {
+                    "train/episode_reward": episode_rewards.mean().item(),
+                    "train/episode_length": episode_length.sum().item()
+                    / len(episode_length),
+                }
+            )
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply examples/dqn/dqn_cartpole.py:128]: I can try to simplify it but I kind of like that most of the logic is here, in a ~200 lines script

File: examples/dqn/dqn_cartpole.py (hunk 10 above)
Comment (albertbou92) [reply examples/dqn/dqn_cartpole.py:124]: loss_td only contains "loss"

File: examples/dqn/dqn_cartpole.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                eval_start = time.time()
+                test_rewards = eval_model(model, test_env, num_test_episodes)
+                eval_time = time.time() - eval_start
+                log_info.update(
+                    {
+                        "eval/reward": test_rewards,
+                        "eval/eval_time": eval_time,
+                    }
+                )
+                model.train()
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply examples/dqn/dqn_cartpole.py]: Same as above, I like seeing the logic, specially in a short script, but I am fine creating function to isolate the whole evaluation part if is preferred

File: examples/dqn/dqn_cartpole.py (hunk 10 above)
Comment (vmoens) [reply examples/dqn/dqn_cartpole.py:124]: haha ok then it's both concise and complete :)

File: setup.py (hunk 7 above)
Comment (albertbou92) [reply setup.py:189]: mm I don't remember adding that, will revert it

File: README.md (hunk 8 above)
Comment (albertbou92) [reply README.md:493]: yes let's just link the pared directory and remove rainbow for now

**Comment by vmoens** [comment]
@albertbou92 the example CI is failing bc we did not update the runs there
